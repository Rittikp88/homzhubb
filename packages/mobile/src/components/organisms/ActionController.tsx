import React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { RecordAssetSelectors } from '@homzhub/common/src/modules/recordAsset/selectors';
import LeaseTermController from '@homzhub/mobile/src/components/organisms/LeaseTermController';
import { SaleTermController } from '@homzhub/mobile/src/components/organisms/SaleTermController';
import { ManageTermController } from '@homzhub/mobile/src/components/organisms/ManageTermController';
import { Country } from '@homzhub/common/src/domain/models/CountryCode';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { AssetGroupTypes } from '@homzhub/common/src/constants/AssetGroup';
import { FurnishingTypes } from '@homzhub/common/src/constants/Terms';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IState } from '@homzhub/common/src/modules/interfaces';

interface IProps {
  typeOfPlan: TypeOfPlan;
  isSplitAsUnits: boolean;
  onNextStep: () => void;
  country: Country;
  assetGroupType: string;
  furnishing: FurnishingTypes;
  lastVisitedStep: ILastVisitedStep;
}

interface IDispatchProps {
  setTermId: (termId: number) => void;
  getMaintenanceUnits: () => void;
}

interface IStateProps {
  currentTermId: number;
  currentAssetId: number;
}

type Props = IStateProps & IDispatchProps & IProps;

// TODO (Aditya): Optimize, look at un-required props drilling
class ActionController extends React.PureComponent<Props, {}> {
  public componentDidMount = (): void => {
    const { getMaintenanceUnits, assetGroupType } = this.props;
    if (assetGroupType === AssetGroupTypes.COM) {
      getMaintenanceUnits();
    }
  };

  public render = (): React.ReactNode => {
    const { onNextStep } = this.props;
    const {
      typeOfPlan,
      currentTermId,
      setTermId,
      currentAssetId,
      country: { currencies },
      assetGroupType,
      lastVisitedStep,
      furnishing,
    } = this.props;
    return (
      <>
        {typeOfPlan === TypeOfPlan.SELL && (
          <SaleTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.RENT && (
          <LeaseTermController
            currentAssetId={currentAssetId}
            currentTermId={currentTermId}
            assetGroupType={assetGroupType}
            setTermId={setTermId}
            furnishing={furnishing}
            lastVisitedStep={lastVisitedStep}
            currencyData={currencies[0]}
            onNextStep={onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.MANAGE && (
          <ManageTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
      </>
    );
  };

  private onNextStep = async (type: TypeOfPlan): Promise<void> => {
    const { onNextStep, lastVisitedStep, currentAssetId } = this.props;

    const last_visited_step = {
      ...lastVisitedStep,
      listing: {
        ...lastVisitedStep.listing,
        type,
        is_listing_created: true,
      },
    };
    await AssetRepository.updateAsset(currentAssetId, { last_visited_step });
    onNextStep();
  };
}

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setTermId, getMaintenanceUnits } = RecordAssetActions;
  return bindActionCreators({ setTermId, getMaintenanceUnits }, dispatch);
};

const mapStateToProps = (state: IState): IStateProps => {
  const { getCurrentTermId, getCurrentAssetId } = RecordAssetSelectors;
  return {
    currentTermId: getCurrentTermId(state),
    currentAssetId: getCurrentAssetId(state),
  };
};

const HOC = connect(mapStateToProps, mapDispatchToProps)(ActionController);
export { HOC as ActionController };
