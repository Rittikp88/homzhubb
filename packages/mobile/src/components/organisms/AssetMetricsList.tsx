import React from 'react';
import { FlatList, StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { images } from '@homzhub/common/src/assets/images';
import { Label, Text, AssetMetrics, Image } from '@homzhub/common/src/components';
import { Miscellaneous } from '@homzhub/common/src/domain/models/AssetMetrics';

interface IProps {
  assetCount?: number;
  subscription: string;
  data: Miscellaneous[];
  containerStyle?: StyleProp<ViewStyle>;
  isPortfolio?: boolean;
  onPlusIconClicked?: () => void;
}

const AssetMetricsList = (props: IProps): React.ReactElement => {
  const { assetCount = 0, subscription, data, isPortfolio = false, onPlusIconClicked, containerStyle } = props;
  const { t } = useTranslation();

  const renderKeyExtractor = (item: Miscellaneous, index: number): string => {
    return `${item.name}-${index}`;
  };

  const bubblePlusIcon = (): void => {
    if (onPlusIconClicked) {
      onPlusIconClicked();
    }
  };

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={styles.heading}>
        <View style={styles.property}>
          <View style={styles.logo}>
            <Image source={images.homzhubDashboard} />
          </View>
          <View>
            <Text type="regular" textType="bold" style={styles.assetCount}>
              {assetCount}
            </Text>
            <View style={styles.propertiesRow}>
              <Text type="small" textType="semiBold" style={styles.propertyText}>
                {t('common:properties')}
              </Text>
              <Icon name={icons.roundFilled} color={theme.colors.darkTint7} size={6} style={styles.circleIcon} />
              <Label type="large" textType="regular">
                {subscription}
              </Label>
            </View>
          </View>
        </View>
        {isPortfolio && (
          <View style={styles.plusIcon}>
            <Icon name={icons.plus} color={theme.colors.primaryColor} size={40} onPress={bubblePlusIcon} />
          </View>
        )}
      </View>
      <View style={styles.assetMetrics}>
        <FlatList
          data={data}
          horizontal
          showsHorizontalScrollIndicator={false}
          renderItem={({ item }: { item: Miscellaneous }): React.ReactElement => {
            const {
              label,
              count,
              colorGradient: { hexColorA, hexColorB, location },
            } = item;
            return (
              <AssetMetrics header={label} value={count} colorA={hexColorA} colorB={hexColorB} location={location} />
            );
          }}
          keyExtractor={renderKeyExtractor}
        />
      </View>
    </View>
  );
};

export { AssetMetricsList };

const styles = StyleSheet.create({
  container: {
    flex: 0,
    backgroundColor: theme.colors.white,
    borderRadius: 4,
    minHeight: 190,
  },
  heading: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    alignItems: 'center',
  },
  property: {
    flexDirection: 'row',
    paddingVertical: 10,
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  logo: {
    flex: 0,
    paddingHorizontal: 15,
  },
  assetCount: {
    color: theme.colors.darkTint1,
  },
  propertyText: {
    color: theme.colors.darkTint4,
    paddingVertical: 5,
  },
  assetMetrics: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    flex: 1,
  },
  plusIcon: {
    backgroundColor: theme.colors.lightGrayishBlue,
    marginRight: 12,
  },
  propertiesRow: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  circleIcon: {
    paddingHorizontal: 8,
    paddingTop: 5,
  },
});
