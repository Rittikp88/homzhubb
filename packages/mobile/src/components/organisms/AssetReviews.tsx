import React, { memo, useEffect, useState } from 'react';
import { View, StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { theme } from '@homzhub/common/src/styles/theme';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { AssetReviewCard } from '@homzhub/mobile/src/components/molecules/AssetReviewCard';
import { AssetReviewsSummary } from '@homzhub/mobile/src/components/molecules/AssetReviewsSummary';
import { AssetReview } from '@homzhub/common/src/domain/models/AssetReview';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { icons } from '@homzhub/common/src/assets/icon';

interface IProps {
  saleListingId: number | null;
  leaseListingId: number | null;
}

const AssetReviews = (props: IProps): React.ReactElement => {
  const { saleListingId, leaseListingId } = props;
  const [reviews, setReviews] = useState<AssetReview[]>([]);
  const [reviewSummary, setReviewSummary] = useState<AssetReview | null>(null);

  const { t } = useTranslation(LocaleConstants.namespacesKey.property);

  useEffect(() => {
    if (!saleListingId && !leaseListingId) {
      return;
    }

    try {
      AssetRepository.getListingReviews({
        ...(leaseListingId && { lease_listing: leaseListingId }),
        ...(saleListingId && { sale_listing: saleListingId }),
      }).then((response) => {
        setReviews(response);
      });
      AssetRepository.getListingReviewsSummary({
        ...(leaseListingId && { lease_listing: leaseListingId }),
        ...(saleListingId && { sale_listing: saleListingId }),
      }).then((response) => {
        setReviewSummary(response);
      });
    } catch (e) {
      AlertHelper.error({ message: t('common:genericErrorMessage') });
    }
  }, []);

  if (reviews.length <= 0 && !reviewSummary) {
    return <EmptyState title={t('property:noReview')} icon={icons.reviews} />;
  }

  return (
    <View style={styles.container}>
      {reviewSummary && <AssetReviewsSummary reviewSummary={reviewSummary} containerStyle={styles.content} />}
      {reviews.length > 0 && (
        <Label type="large" textType="semiBold" style={styles.popularWith}>
          {t('common:popularWith')}
        </Label>
      )}
      {reviews.map((review: AssetReview) => (
        <AssetReviewCard
          key={review.id}
          description={review.description}
          reviewedAt={review.postedAt}
          reviewedBy={review.reviewedBy}
          isReported={review.isReported}
          overallRating={review.rating}
          pillars={review.pillarRatings}
        />
      ))}
    </View>
  );
};

const memoizedComponent = memo(AssetReviews);
export { memoizedComponent as AssetReviews };

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.colors.white,
    minHeight: 200,
  },
  content: {
    marginHorizontal: 16,
  },
  popularWith: {
    color: theme.colors.darkTint3,
    margin: 16,
  },
});
