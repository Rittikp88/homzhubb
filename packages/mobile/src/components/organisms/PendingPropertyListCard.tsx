import React, { Component } from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { PropertyUtils } from '@homzhub/common/src/utils/PropertyUtils';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Button, Label, Text } from '@homzhub/common/src/components';
import { PropertyAmenities, PropertyAddressCountry, ProgressBar, ShieldGroup } from '@homzhub/mobile/src/components';
import { PropertyReviewCard } from '@homzhub/mobile/src/components/molecules/PropertyReviewCard';
import { Asset, LastVisitedStep, PropertyStatus } from '@homzhub/common/src/domain/models/Asset';
import { IActions, TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { IAmenitiesIcons } from '@homzhub/common/src/domain/models/Search';

const actionButtons: IActions[] = [
  { id: 1, title: 'Invite Tenant', type: TypeOfPlan.MANAGE },
  { id: 2, title: 'Rent', type: TypeOfPlan.RENT },
  { id: 3, title: 'Sell', type: TypeOfPlan.SELL },
];

interface IProps {
  onPressComplete: (assetId: number) => void;
  onSelectAction: (payload: IActions, assetId: number) => void;
}

interface IState {
  currentPropertyIndex: number;
  data: Asset[];
  isActionsVisible: boolean;
}

type Props = IProps & WithTranslation;

export class PendingPropertyListCard extends Component<Props, IState> {
  public state = {
    currentPropertyIndex: 0,
    data: [],
    isActionsVisible: false,
  };

  public componentDidMount = async (): Promise<void> => {
    await this.getPendingProperties();
  };

  public render(): React.ReactNode {
    const { currentPropertyIndex, data } = this.state;
    const { t } = this.props;
    const currentProperty = data[currentPropertyIndex];
    const total = data.length;

    return (
      <View style={styles.container}>
        <View style={styles.headingView}>
          <View style={styles.headingContent}>
            <Icon name={icons.warning} size={16} />
            <Text type="small" textType="semiBold" style={styles.label}>
              {t('pendingProperties', { total })}
            </Text>
          </View>
          <View style={styles.headingContent}>
            <TouchableOpacity style={styles.iconStyle}>
              <Icon
                name={icons.leftArrow}
                size={16}
                onPress={this.handlePrevious}
                color={currentPropertyIndex === 0 ? theme.colors.darkTint4 : theme.colors.primaryColor}
                testID="icnPrevious"
              />
            </TouchableOpacity>
            <TouchableOpacity style={styles.iconStyle} disabled={currentPropertyIndex === total - 1}>
              <Icon
                name={icons.rightArrow}
                size={16}
                onPress={this.handleNext}
                color={currentPropertyIndex === total - 1 ? theme.colors.darkTint4 : theme.colors.primaryColor}
                testID="icnNext"
              />
            </TouchableOpacity>
          </View>
        </View>
        {currentProperty && this.renderCardItem(currentProperty)}
      </View>
    );
  }

  private renderCardItem = (item: Asset): React.ReactElement => {
    const {
      id,
      spaces,
      assetType: { name },
      projectName,
      unitNumber,
      carpetArea,
      carpetAreaUnit,
      floorNumber,
      blockNumber,
      address,
      lastVisitedStep,
      verifications: { description },
    } = item;
    const amenitiesData: IAmenitiesIcons[] = PropertyUtils.getAmenities(
      spaces,
      floorNumber,
      item.assetGroup.name,
      carpetArea,
      carpetAreaUnit?.title ?? '',
      true
    );

    return (
      <View style={styles.cardContainer}>
        <ShieldGroup propertyType={name} propertyTypeStyle={styles.heading} text={description} isInfoRequired />
        <PropertyAddressCountry
          primaryAddress={projectName}
          subAddress={address ?? `${unitNumber ?? ''} ${blockNumber ?? ''}`}
          containerStyle={styles.addressStyle}
        />
        {amenitiesData.length > 0 && (
          <PropertyAmenities data={amenitiesData} direction="row" containerStyle={styles.amenitiesContainer} />
        )}
        {this.renderButtonsView(id, lastVisitedStep)}
      </View>
    );
  };

  private renderButtonsView = (id: number, lastVisitedStep: LastVisitedStep): React.ReactElement => {
    const { t, onPressComplete, onSelectAction } = this.props;
    const { isActionsVisible } = this.state;
    const {
      assetCreation: { percentage },
      listing: { type },
      isVerificationRequired,
      isListingRequired,
      isPropertyReady,
    } = lastVisitedStep;
    const plan = actionButtons.find((item) => item.type === type);
    let onVerifyProperty;
    if (plan) {
      onVerifyProperty = (): void => onSelectAction(plan, id);
    }
    return (
      <>
        {percentage < 100 && (
          <>
            <ProgressBar progress={percentage} width={theme.viewport.width > 400 ? 350 : 310} />
            <Button
              type="primary"
              title={t('completeDetails')}
              containerStyle={styles.buttonStyle}
              onPress={(): void => onPressComplete(id)}
            />
          </>
        )}
        {isListingRequired && (
          <Button
            type="primary"
            title={t('takeActions')}
            iconSize={24}
            iconColor={theme.colors.blue}
            icon={isActionsVisible ? icons.upArrow : icons.downArrow}
            containerStyle={styles.actionButton}
            titleStyle={styles.actionButtonTitle}
            onPress={this.handleActions}
          />
        )}
        {isListingRequired && isActionsVisible && (
          <>
            {actionButtons.map((item) => {
              return (
                <Button
                  key={item.id}
                  type="secondary"
                  title={item.title}
                  containerStyle={styles.buttonStyle}
                  onPress={(): void => onSelectAction(item, id)}
                />
              );
            })}
          </>
        )}
        {isVerificationRequired && (
          <Button
            type="primary"
            title={t('completeVerification')}
            containerStyle={styles.buttonStyle}
            onPress={onVerifyProperty}
          />
        )}
        {isPropertyReady && <PropertyReviewCard />}
        {percentage < 100 && (
          <Label type="regular" style={styles.infoText}>
            {t('completeProperty')}
          </Label>
        )}
      </>
    );
  };

  private getPendingProperties = async (): Promise<void> => {
    try {
      const response: Asset[] = await AssetRepository.getPropertiesByStatus(PropertyStatus.PENDING);
      this.setState({ data: response });
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private handleNext = (): void => {
    const { currentPropertyIndex, data } = this.state;
    if (data.length !== currentPropertyIndex + 1) {
      this.setState({ currentPropertyIndex: currentPropertyIndex + 1 });
    }
  };

  private handlePrevious = (): void => {
    const { currentPropertyIndex } = this.state;
    if (currentPropertyIndex !== 0) {
      this.setState({ currentPropertyIndex: currentPropertyIndex - 1 });
    }
  };

  private handleActions = (): void => {
    this.setState((prevState) => {
      return {
        isActionsVisible: !prevState.isActionsVisible,
      };
    });
  };
}

export default withTranslation(LocaleConstants.namespacesKey.assetDashboard)(PendingPropertyListCard);

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.colors.white,
    borderRadius: 4,
    marginTop: 12,
    padding: 16,
  },
  headingView: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  headingContent: {
    flexDirection: 'row',
    alignItems: 'center',
    height: 40,
  },
  label: {
    color: theme.colors.darkTint1,
    marginLeft: 10,
  },
  iconStyle: {
    width: PlatformUtils.isIOS() ? 32 : 38,
    height: PlatformUtils.isIOS() ? 30 : 35,
    backgroundColor: theme.colors.background,
    alignItems: 'center',
    justifyContent: 'center',
    marginRight: 6,
  },
  cardContainer: {
    marginVertical: 12,
  },
  heading: {
    color: theme.colors.darkTint3,
  },
  amenitiesContainer: {
    marginBottom: 20,
  },
  buttonStyle: {
    flex: 0,
    marginBottom: 5,
    marginTop: 18,
  },
  actionButton: {
    flex: 0,
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 5,
    marginTop: 18,
    backgroundColor: theme.colors.blueOpacity,
  },
  actionButtonTitle: {
    color: theme.colors.blue,
    marginHorizontal: 6,
  },
  infoText: {
    color: theme.colors.darkTint7,
    textAlign: 'center',
  },
  addressStyle: {
    marginBottom: 6,
  },
});
