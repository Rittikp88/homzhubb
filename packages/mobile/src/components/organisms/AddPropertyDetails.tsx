import React, { ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button, FormTextInput, FurnishingSelection, Text, WithShadowView } from '@homzhub/common/src/components';
import { AssetDescriptionForm } from '@homzhub/mobile/src/components/molecules/AssetDescriptionForm';
import { PropertySpaces } from '@homzhub/mobile/src/components/organisms/PropertySpaces';
import { AssetListingSection } from '@homzhub/mobile/src/components/HOC/AssetListingSection';
import { AssetDescriptionDropdownValues } from '@homzhub/common/src/domain/models/AssetDescriptionForm';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { ISpaceCount, SpaceType } from '@homzhub/common/src/domain/models/AssetGroup';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IUpdateAssetParams } from '@homzhub/common/src/domain/repositories/interfaces';

interface IDescriptionForm {
  carpetArea?: number;
  areaUnit?: number;
  buildingGrade?: string;
  facing?: string;
  flooringType?: number;
  yearOfConstruction?: number;
  totalFloors?: number;
  onFloorNumber?: number;
}

interface IFurnishingForm {
  furnishingDetails?: string;
  furnishingType?: string;
}

interface IOwnProps extends WithTranslation {
  assetId: number;
  assetDetails: Asset | null;
  spaceTypes: SpaceType[];
  handleNextStep: () => void;
  lastVisitedStep: ILastVisitedStep;
}

interface IOwnState {
  spacesForm: ISpaceCount[];
  descriptionForm: IDescriptionForm;
  furnishingForm: IFurnishingForm;
  descriptionDropdownValues: AssetDescriptionDropdownValues | null;
}

class AddPropertyDetails extends React.PureComponent<IOwnProps, IOwnState> {
  constructor(props: IOwnProps) {
    super(props);
    const { assetDetails } = this.props;

    this.state = {
      spacesForm: [],
      descriptionForm: {
        carpetArea: (assetDetails && assetDetails.carpetArea) || undefined,
        areaUnit: (assetDetails && assetDetails.carpetAreaUnit && assetDetails.carpetAreaUnit.id) || 1,
        buildingGrade: '',
        facing: (assetDetails && assetDetails.facing) || undefined,
        flooringType: (assetDetails && assetDetails.floorType) || undefined,
        yearOfConstruction: (assetDetails && assetDetails.construction_Year) || undefined,
        totalFloors: (assetDetails && assetDetails.totalFloors) || 0,
        onFloorNumber: (assetDetails && assetDetails.floorNumber) || 0,
      },
      furnishingForm: {
        furnishingDetails: (assetDetails && assetDetails.furnishingDescription) || undefined,
        furnishingType: (assetDetails && assetDetails.furnishing) || undefined,
      },
      descriptionDropdownValues: null,
    };
  }

  public async componentDidMount(): Promise<void> {
    await this.getDescriptionDropdownValues();
  }

  public render(): ReactElement {
    const { spaceTypes, t } = this.props;
    const { descriptionForm, furnishingForm, descriptionDropdownValues } = this.state;

    return (
      <>
        <Formik
          onSubmit={this.onSubmit}
          initialValues={{ ...descriptionForm, ...furnishingForm }}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<FormikValues>): React.ReactNode => {
            return (
              <>
                <View style={styles.containerStyle}>
                  <>
                    <Text style={styles.headingStyle} type="small">
                      {t('property:spacesText')}
                    </Text>
                    <PropertySpaces onChange={this.handleSpaceFormChange} spacesTypes={spaceTypes} />
                  </>

                  {descriptionDropdownValues && (
                    <AssetDescriptionForm dropDownOptions={descriptionDropdownValues} formProps={formProps} />
                  )}

                  {descriptionDropdownValues && this.renderFurnishingFields(formProps)}
                </View>

                <WithShadowView>
                  <Button
                    type="primary"
                    title={t('common:continue')}
                    containerStyle={styles.buttonStyle}
                    onPress={(): Promise<void> => this.onSubmit(formProps.values)}
                  />
                </WithShadowView>
              </>
            );
          }}
        </Formik>
      </>
    );
  }

  private renderFurnishingFields = (formProps: FormikProps<FormikValues>): ReactElement => {
    const { t } = this.props;
    const onFurnishingChange = (value: string): void => this.setFurnishingStatus(formProps, value);

    return (
      <AssetListingSection title={t('property:furnishing')} containerStyles={styles.furnishingStyle}>
        <>
          <FurnishingSelection
            titleHidden
            value={formProps.values.furnishingType}
            onFurnishingChange={onFurnishingChange}
          />
          <FormTextInput
            style={styles.furnishingFieldStyle}
            name="furnishingDetails"
            label={t('property:furnishingDetails')}
            maxLength={400}
            inputType="default"
            formProps={formProps}
          />
        </>
      </AssetListingSection>
    );
  };

  private onSubmit = async (values: FormikValues): Promise<void> => {
    const {
      areaUnit,
      carpetArea,
      facing,
      flooringType,
      furnishingDetails,
      onFloorNumber,
      totalFloors,
      yearOfConstruction,
      furnishingType,
    } = values;
    const { spacesForm } = this.state;
    const { handleNextStep, assetId, lastVisitedStep } = this.props;

    const sanitizedSpaces = spacesForm
      .filter((item) => item && item.description !== '')
      .map((space) => {
        return {
          space_type: space.space_type,
          count: space.count,
          ...(space.description && { description: space.description }),
        };
      });

    const payload: IUpdateAssetParams = {
      carpet_area: carpetArea,
      carpet_area_unit: areaUnit,
      facing,
      floor_type: flooringType,
      furnishing_description: furnishingDetails,
      floor_number: onFloorNumber,
      total_floors: totalFloors,
      construction_year: yearOfConstruction,
      furnishing: furnishingType,
      spaces: sanitizedSpaces,
      last_visited_step: {
        ...lastVisitedStep,
        asset_creation: {
          ...lastVisitedStep.asset_creation,
          is_details_done: true,
          total_step: 4,
        },
      },
    };

    try {
      await AssetRepository.updateAsset(assetId, payload);
      handleNextStep();
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private setFurnishingStatus = (formProps: FormikProps<FormikValues>, furnishingType: string): void => {
    formProps.setFieldValue('furnishingType', furnishingType);
  };

  private handleSpaceFormChange = (id: number, count: number, description?: string): void => {
    const { spacesForm } = this.state;

    spacesForm[id] = { space_type: id, count, description };
  };

  private formSchema = (): yup.ObjectSchema<IDescriptionForm & IFurnishingForm> => {
    return yup.object().shape({
      carpetArea: yup.number().optional(),
      areaUnit: yup.number().optional(),
      buildingGrade: yup.string().optional(),
      facing: yup.string().optional(),
      flooringType: yup.number().optional(),
      yearOfConstruction: yup.number().optional(),
      totalFloors: yup.number().optional(),
      onFloorNumber: yup.number().optional(),
      furnishingType: yup.string().optional(),
      furnishingDetails: yup.string().optional(),
    });
  };

  private getDescriptionDropdownValues = async (): Promise<void> => {
    try {
      const response: AssetDescriptionDropdownValues = await AssetRepository.getAssetDescriptionDropdownValues();

      this.setState({
        descriptionDropdownValues: response,
        furnishingForm: {
          furnishingType: response.furnishingStatus[0].name,
        },
      });
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };
}

const addPropertyDetails = withTranslation()(AddPropertyDetails);
export { addPropertyDetails as AddPropertyDetails };

const styles = StyleSheet.create({
  containerStyle: {
    paddingHorizontal: theme.layout.screenPadding,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  furnishingStyle: {
    marginTop: 16,
  },
  headingStyle: {
    marginTop: 16,
    paddingVertical: 16,
    paddingLeft: 16,
    backgroundColor: theme.colors.moreSeparator,
  },
  furnishingFieldStyle: {
    height: 85,
    textAlignVertical: 'top',
    paddingTop: 10,
  },
});
