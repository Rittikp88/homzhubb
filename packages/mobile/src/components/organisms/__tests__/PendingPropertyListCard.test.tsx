import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PendingPropertyListCard } from '@homzhub/mobile/src/components/organisms/PendingPropertyListCard';
import { mockAsset } from '@homzhub/common/src/mocks/AssetDescription';

let props: any;
let wrapper: ShallowWrapper;

describe('PendingPropertyListCard', () => {
  wrapper = shallow(<PendingPropertyListCard {...props} t={(key: string): string => key} />);
  wrapper.setState({ data: [mockAsset] });

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
