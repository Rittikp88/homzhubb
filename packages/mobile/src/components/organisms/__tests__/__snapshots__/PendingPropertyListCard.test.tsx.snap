// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingPropertyListCard should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 4,
      "marginTop": 12,
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
        }
      }
    >
      <Icon
        allowFontScaling={false}
        name="warning"
        size={16}
      />
      <Text
        style={
          Object {
            "color": "#001B36",
            "marginLeft": 10,
          }
        }
        textType="semiBold"
        type="small"
      >
        pendingProperties
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginVertical": 12,
      }
    }
  >
    <ShieldGroup
      isInfoRequired={true}
      propertyType="Apartment / Condo"
      propertyTypeStyle={
        Object {
          "color": "#33495E",
        }
      }
      text="It was popularised in the 1960s with the release of Letraset sheets."
    />
    <PropertyAddressCountry
      containerStyle={
        Object {
          "marginBottom": 6,
        }
      }
      subAddress=" "
    />
    <PropertyAmenities
      containerStyle={
        Object {
          "marginBottom": 20,
        }
      }
      data={
        Array [
          Object {
            "icon": "furnishing",
            "iconColor": "#33495E",
            "iconSize": 20,
            "label": "None",
          },
          Object {
            "icon": "open-parking",
            "iconColor": "#33495E",
            "iconSize": 20,
            "label": "-",
          },
          Object {
            "icon": "area",
            "iconColor": "#33495E",
            "iconSize": 20,
            "label": "-",
          },
        ]
      }
      direction="row"
    />
  </View>
</View>
`;
